File: bot-api/dotnet/test/src/TeamMessageBatchTest.cs
using NUnit.Framework;
using Robocode.TankRoyale.BotApi;
using Robocode.TankRoyale.BotApi.Internal;

namespace Robocode.TankRoyale.BotApi.Tests;

[TestFixture]
[Category("TR-API-TCK-022")]
public class TeamMessageBatchTest
{
    [Test]
    public void BatchPayloadKeepsOrderAndUsesReservedMarker()
    {
        var batch = new TeamMessageBatch(new[] { "first", "second" });
        var payload = TeamMessageBatchCodec.Encode(batch);
        var items = TeamMessageBatchCodec.DecodeItems(payload);
        Assert.That(items.Count, Is.EqualTo(2));
        Assert.That((string)items[0]["message"], Is.EqualTo("\"first\""));
        Assert.That((string)items[1]["message"], Is.EqualTo("\"second\""));
        Assert.That(TeamMessageBatchCodec.MessageType, Is.EqualTo("team-message-batch-v1"));
    }

    [Test]
    public void BatchRequiresAtLeastOneNonNullPayload()
    {
        Assert.Throws<System.ArgumentException>(() => new TeamMessageBatch(System.Array.Empty<object>()));
        Assert.Throws<System.ArgumentException>(() => new TeamMessageBatch(new object[] { "ok", null }));
    }

    [Test]
    public void LogicalItemLimitAccepts128AndRejects129()
    {
        Assert.DoesNotThrow(() => IntentValidator.ValidateLogicalTeamMessageCount(128));
        Assert.Throws<BotException>(() => IntentValidator.ValidateLogicalTeamMessageCount(129));
    }
}

File: bot-api/typescript/test/TeamMessageBatch.test.ts
import { describe, expect, it } from "vitest";
import { TeamMessageBatch } from "../src/TeamMessageBatch.js";
import { Constants } from "../src/Constants.js";
import { IntentValidator } from "../src/internal/intentValidator.js";
import { BotException } from "../src/BotException.js";

describe("TR-API-TCK-022: TeamMessageBatch", () => {
  it("keeps order and exposes the shared logical item limit", () => {
    expect(new TeamMessageBatch(["first", "second"]).messages).toEqual(["first", "second"]);
    expect(Constants.MAX_LOGICAL_TEAM_MESSAGES_PER_TURN).toBe(128);
    expect(() => IntentValidator.validateLogicalTeamMessageCount(128)).not.toThrow();
    expect(() => IntentValidator.validateLogicalTeamMessageCount(129)).toThrow(BotException);
  });

  it("requires at least one non-null payload", () => {
    expect(() => new TeamMessageBatch([])).toThrow("at least one message");
    expect(() => new TeamMessageBatch(["ok", null])).toThrow("cannot contain null");
  });
});
